format/src/main/scala/org/scalafmt/internal/SyntaxTokens.scala

```diff
@@ -282,6 +268,114 @@ object SyntaxTokens {
     def tokensWith: KwWith = tree.findBetween[KwWith](_.lhs, _.rhs).get
   }
 
+  // == Defn ==
+  implicit class XtensionDefnClassSyntax(val tree: Defn.Class) extends AnyVal {
+    def tokensClass: KwClass = tree.find[KwClass].get
+
+    // tparams
+    def tokensLeftBracket: Option[LeftBracket] =
+      tree.tparams.headOption
+        .map(tparam => tree.findBetween[LeftBracket](_.name, _ => tparam).get)
+    def tokensRightBracket: Option[RightBracket] =
+      tree.tparams.lastOption.map(
+        tparam =>
+          tree.ctor.paramss match {
+            case Nil => tree.findAfter[RightBracket](_ => tparam).get
+            case xs :: _ =>
+              xs match {
+                case Nil => tree.findAfter[RightBracket](_ => tparam).get
+                case x :: _ =>
+                  tree.findBetween[RightBracket](_ => tparam, _ => x).get
+              }
+        }
+      )
+
+    def tokensCommaTparams: List[Comma] =
+      commaSeparated(tree)(_.tparams)
+
+    def tokensCommaCtor: List[List[Comma]] =
+      tree.ctor.paramss.map(commaSeparated0(tree))
+
+    def tokensImplicit: Option[KwImplicit] = {
+      val paramss = tree.ctor.paramss
+      if (paramss.isEmpty) None
+      else {
+        val rparamss = paramss.reverse
+        rparamss.head match {
+          case right :: _ if right.mods.exists(_.is[Mod.Implicit]) =>
+            val list = TokenList(tree.tokens)
```

Comment:
we should replace TokenList.leading/trailing with stateless extension methods on `Tokens` using binary search. Constructing a TokenList is fairly expensive because it compute lookup maps

Reply:
is this still true with binarySearch ?

Reply:
nope, hadn't seen you removed tok2idx. I still think `TokenList` should be replaced with extension methods